client/core/servercontroller.cpp

```diff
@@ -856,3 +862,45 @@ QString ServerController::replaceVars(const QString &script, const Vars &vars)
     //qDebug().noquote() << script;
     return s;
 }
+
+ErrorCode ServerController::isServerPortBusy(const ServerCredentials &credentials, DockerContainer container)
+{
+    QString stdOut;
+    auto cbReadStdOut = [&](const QString &data, QSharedPointer<QSsh::SshRemoteProcess> proc) {
+        stdOut += data + "\n";
+    };
+    auto cbReadStdErr = [&](const QString &data, QSharedPointer<QSsh::SshRemoteProcess> ) {
+        stdOut += data + "\n";
+    };
+
+    QString script = QString("sudo lsof -i -P -n | grep -E ");
+
+    switch (container) {
+    case DockerContainer::WireGuard: {
+        script.append("':$WIREGUARD_SERVER_PORT'");
+        break;
+    }
+    case DockerContainer::OpenVpn:{
+        script.append("':$OPENVPN_PORT'");
+        break;
+    }
+    case DockerContainer::ShadowSocks:{
+        script.append("':$SHADOWSOCKS_SERVER_PORT'");
+        break;
+    }
+    case DockerContainer::Cloak:{
+        script.append("':$CLOAK_SERVER_PORT'");
+        break;
+    }
+    default:
+        return ErrorCode::NoError;
+    }
+
```

Comment:
IKEv2, DNS, SFTP containers missing, also missing protocol TCP/UDP detection.
You need to use "port" and "transport_proto" that is already present in each container config.
For containers like IKEv2 where ports are fixed, and in amnezia's config it's set to 0, I suggest to implement some new function like ContainerProps::fixedPorts(DockerContainer c) which will return list of fixed ports.
When done, you will be able to get rid of this boilerplate, just add two checks - for actual container ports, and for fixed ports. I guess we will implement new containers with fixed ports sometime.